### PR TITLE
Add auto-diff support for `IRSwizzleStore`

### PR DESCRIPTION
**Pull request by saipraveenb25**
- Lower IRSwizzleStore to multiple stores during AD preprocess.
 - Fix typo in `transcribeNonDiffInst`

### CODE REVIEW

File: out.glsl [hunk 1]
```diff
@@ -0,0 +1,49 @@
+#version 450
```
Comment (csyonghe) [inline out.glsl:1]: unrelated?

File: source/slang/slang-ir-autodiff-fwd.cpp [hunk 2]
```diff
@@ -1617,6 +1617,54 @@ void insertTempVarForMutableParams(IRModule* module, IRFunc* func)
     }
 }
 
+bool isResourcePointer(IRInst* ptrInst)
+{
+    if (auto getElementPtr = as<IRGetElementPtr>(ptrInst))
```
Comment (csyonghe) [inline source/slang/slang-ir-autodiff-fwd.cpp:1622]: You can use `getRootAddr` here.

**Review by csyonghe — APPROVED** [review]

File: source/slang/slang-ir-autodiff-fwd.cpp [hunk 3]
```diff
@@ -1617,6 +1617,50 @@ void insertTempVarForMutableParams(IRModule* module, IRFunc* func)
     }
 }
 
+bool isLocalPointer(IRInst* ptrInst)
+{
+    // If it's not a local var or a function parameter, then it's probably 
+    // referencing something outside the function scope.
+    // 
+    return as<IRVar>(getRootAddr(ptrInst)) || as<IRParam>(getRootAddr(ptrInst));
```
Comment (csyonghe) [inline source/slang/slang-ir-autodiff-fwd.cpp:1625]: You may want to call getRootAddr once.

File: source/slang/slang-ir-autodiff-fwd.cpp [hunk 4]
```diff
@@ -1626,6 +1670,8 @@ SlangResult ForwardDiffTranscriber::prepareFuncForForwardDiff(IRFunc* func)
     
     initializeLocalVariables(autoDiffSharedContext->moduleInst->getModule(), func);
 
+    lowerSwizzledStores(autoDiffSharedContext->moduleInst->getModule(), func);
```
Comment (csyonghe) [inline source/slang/slang-ir-autodiff-fwd.cpp:1674]: It will be beneficial to move this to cover all functions and not just the AD ones. We can run the pass again at the end of linkAndOptimizeIR, before hitting emit. A good place to call it might be after `finalizeSpecialization`.

**Review by csyonghe — DISMISSED** [review]

File: source/slang/slang-ir-autodiff-fwd.cpp (hunk 4 above)
Comment (csyonghe) [reply source/slang/slang-ir-autodiff-fwd.cpp:1674]: A pitfall though is if I have a function taking an `__ref` parameter and `swizzleStore` it, the function will get inlined later, and we probably don't want to lower the swizzled store on that parameter just yet.

Comment (saipraveenb25) [reply source/slang/slang-ir-autodiff-fwd.cpp:1674]: Hmm.. should we run the pass again after those methods are inlined?

Comment (csyonghe) [reply source/slang/slang-ir-autodiff-fwd.cpp:1674]: That sort of defeats the purpose of the pass, which is removing the concern of `swizzledStore` throughout the codebase...

Comment (csyonghe) [reply source/slang/slang-ir-autodiff-fwd.cpp:1674]: I guess its fine to just run the pass for AD then...

Comment (csyonghe) [reply source/slang/slang-ir-autodiff-fwd.cpp:1674]: It looks like the right choice is to implement a targeted optimization in image load/store lowering pass to detect and combine multiple stores, instead of trying to preserve swizzledStore all over the place.

**Review by csyonghe — APPROVED** [review]
